addons/stock_account/models/stock.py

```diff
@@ -75,95 +76,218 @@ class StockMove(models.Model):
+        next_moves = self.search(all_domain, order='date, id')
         if start_move:
             last_cumulated_value = start_move.cumulated_value
             last_done_qty_available = start_move.last_done_qty
         else:
             last_cumulated_value = 0.0
             last_done_qty_available = 0.0
-        if self.product_id.cost_method == 'average':
-            for move in next_moves:
-                if move.location_id.usage in ('internal', 'transit') and move.location_dest_id.usage not in ('internal', 'transit'):
-                    if last_done_qty_available:
-                        move.value = - ((last_cumulated_value / last_done_qty_available) * move.product_qty)
-                    last_done_qty_available -= move.product_qty
-                else:
-                    last_done_qty_available += move.product_qty
-                move.cumulated_value = last_cumulated_value + move.value
-                last_cumulated_value = move.cumulated_value
-                move.last_done_qty = last_done_qty_available
-        
-        
-        # FIFO: needs dict with qty_remaining to replay algorithm
-        
-        
-        # update cumulated_value according to value
-        
-        # update outs according to values
+        for move in next_moves:
+            if move._is_out_move():
+                if last_done_qty_available:
+                    move.value = - (last_cumulated_value * move.product_qty / last_done_qty_available)
+                last_done_qty_available -= move.product_qty
+            else:
+                last_done_qty_available += move.product_qty
+            last_cumulated_value += move.value
+            move.write({'cumulated_value': last_cumulated_value, 
+                        'last_done_qty': last_done_qty_available})
+
+    def replay_fifo(self):
+        """
+            We try to find an outgoing move before this one that did not 
+            have a negative stock when it was done and has a corresponding in move. 
+            By taking the later in moves and out moves, we recalculate the remaining qties and values on the moves
+            Afterwards, we take both in and out moves chronologically and recalculate cumulated value and qty done
+        """
+        out_domain = self._get_out_domain()
+        out_domain += ['|', ('date', '<', self.date),
+                     '&', ('date', '=', self.date), ('id', '<', self.id), 
+                     ('last_done_move_id', '!=', False),
+                     ('last_done_qty', '>=', 0.0)]
+        start_move = self.search(out_domain, limit=1, order='date desc, id desc')
+        # normally, this will be the case as last_done_qty > 0.0
+        use_start_move = start_move and start_move.last_done_move_id and (start_move.last_done_move_id.date < start_move.date or (start_move.last_done_move_id.date == start_move.date and start_move.last_done_move_id.id < start_move.id))
```

Comment:
start_move.last_done_move_id.date < start_move.date excludes negative moves that were compensated